### ISSUE TITLE
[QUESTION] Record List View Permissions based on Team

### ISSUE THREAD

**Issue by blackdrago** [santigarcor/laratrust#306]
- Laravel Version: 5.4
- Laratrust Version: 5.x (not started yet)

### Description:
I'm working on a project that has team-driven access. All access is driven by groups (teams). I've read a lot of examples of how permissions are granted to users/roles for specific models/actions (e.g., "create-contact", "edit-contact", etc.), but I'm not clear how this works with the visibility of records/objects. 

Specifically, if I have a List View of records/objects, I want each user to see only the records they have access to view (read). For this project, access is determined by group membership. If User A and User B are both on Team 1, they can both see/read each other's records, but User Y and User Z (both on Team 2) cannot see User A's and User B's records, and vice-versa.

I was imagining some of a join table with additional params for SQL, and/or some kind of filtering function for models.

It looks like this might be possible by adding teams and then by implementing the Ownable interface so that "own" applies to anyone inside the team, but doing that would mean I'd need to think about ways to handle true user ownership (e.g., only the creating user can modify their own record). 

But even if I did modify ownership to apply to teams, it seems as if the package requires me to instantiate each record as an object and then check it against the user:

```php
$objectList = [];
$theObject = new SomeOwnedObject::all();
if ($user->owns($theObject)) {
    $objectList[] = $theObject;
}
```
This will take up a lot more memory than querying the database with the permissions for each record, something like:
```php
$objectList = SomeOwnedObject::withAccessFor($user);
```
or even
```php
$objectList = SomeOwnedObject::all()->withAccessFor($user);
```
This way, there's no processing power/time spent creating objects that the user doesn't have access to. 

Is this possible in Laratrust? Is there a team ownership element that can be applied without customizing "owns"?

Thank you!

**Comment by santigarcor** [comment]
This is not possible right now, because there's no method implemented to make that. If you want you are more than welcome to add a PR adding that feature.

**Comment by MenezesVinicius** [comment]
Any updates?

**Comment by blackdrago** [comment]
Due to other design requirements, we decided to implement our own private package for our applications... The solution I used for the scenario I described went like this:

- created a table/Model for Access Record, which tracks a given record's creating user, owning group, etc. 
- these Access Records are maintained by an Observer that fires on key events (created, updated, saved, deleted, etc.) for Models identified as Data Restricted (identified via trait mentioned below)
- create a trait for Models that adds a global scope to the Model's boot() method that checks the Access Record against the Current User's group memberships(s) and only includes the Models they have access to see (this way, so long as you use Eloquent models, Data Restriction is built in).

I'm not sure how much of the solution can be adapted to Laratrust. I've been meaning to put together a PR, but keep getting distracted by shiny things.
